### PR TITLE
Support for partition_cols in to_parquet

### PR DESCRIPTION
**Pull request by anjsudh**
- [x] closes [#23283](https://github.com/pandas-dev/pandas/issues/23283)
- [x] tests passed
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR! Please make sure that you always have tests first and foremost with any submission though. 

Once added feel free to ping back and can take another look

File: pandas/io/parquet.py [hunk 1]
```diff
@@ -124,9 +124,15 @@ def write(self, df, path, compression='snappy',
 
         else:
             table = self.api.Table.from_pandas(df, **from_pandas_kwargs)
-            self.api.parquet.write_table(
-                table, path, compression=compression,
-                coerce_timestamps=coerce_timestamps, **kwargs)
+            if 'partition_cols' in kwargs and \
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: style nitpick: we use parenthesis to wrap long lines instead of a backslash.

File: pandas/tests/io/test_parquet.py [hunk 2]
```diff
@@ -474,6 +474,19 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
         check_round_trip(df_compat, pa,
                          path='s3://pandas-test/pyarrow.parquet')
 
+    def test_partition_cols_supported(self, pa, df_full):
+        partition_cols =['bool','int']
+        df = df_full
+        path="a.txt"
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: We use temporary files so that tests don't litter the current directory. Something like

```
    with tm.ensure_clean() as path:
        df.to_parquet(path, engine=pa, ...)
```

File: pandas/tests/io/test_parquet.py [hunk 3]
```diff
@@ -474,6 +474,19 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
...
+        df.to_parquet(path, partition_cols= partition_cols, compression= None)
+        import pyarrow.parquet as pq
+        from pyarrow.filesystem import LocalFileSystem
+        fs = LocalFileSystem.get_instance()
+        dataset = pq.ParquetDataset(path, filesystem=fs, metadata_nthreads=16)
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: is this `metadata_nthreads` necessary?

**Review by TomAugspurger — COMMENTED** [review]
Could you add a release note in 0.24.0 under enhancements? We'll also need to update the docstring for `fname`. IIUC, this will create a directory, rather than a single file, when `partition_cols` is included.

**Comment by TomAugspurger** [comment]
The error in https://travis-ci.org/pandas-dev/pandas/jobs/446310514#L2577 suggests that we'll need to put a minimum pyarrow version to support this behavior. Can you check what version that is, and raise a with an ImportError with a nice error message if the pyarrow is too old?

File: pandas/io/parquet.py [hunk 4]
```diff
@@ -252,7 +258,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
     path : string
-        File path
+        File path ( Will be used as Directory path if
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: Can remove the parenthesis, replace with a new sentence.

Rather than "Directory path", say "`root_path`", which is the name used by `pyarrow.parquet.write_to_dataset`.

File: pandas/io/parquet.py [hunk 5]
```diff
@@ -252,7 +258,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
...
+        'partition_cols' is provided as parameter for 'pyarrow' engine).
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: Typically parameter names like 'partition_cols' are placed in single backticks.

File: pandas/tests/io/test_parquet.py [hunk 6]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
         check_round_trip(df_compat, pa,
                          path='s3://pandas-test/pyarrow.parquet')
 
+    def test_partition_cols_supported(self, pa_ge_070, df_full):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        path = tempfile.mkdtemp()
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: This can use `with tm.ensure_clean() as path:`. Then you don't have to worry about the `shutil.rmtree` and the imports.

File: pandas/tests/io/test_parquet.py [hunk 7]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
...
+        # supported in >= 0.7.0
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: Can remove this comment, since we have the `pa_ge_070` fixture.

File: pandas/tests/io/test_parquet.py [hunk 8]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
...
+        df.to_parquet(path, partition_cols=partition_cols,
+                      compression=None)
+        import pyarrow.parquet as pq
+        dataset = pq.ParquetDataset(path, validate_schema=False)
+        assert len(dataset.pieces) > 0
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: Can we make this assert more precise? Do we know that it will be equal to the number of combinations in the `['bool', 'int']` columns?

File: pandas/tests/io/test_parquet.py [hunk 9]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
...
+        assert len(dataset.partitions.partition_names) > 0
+        assert dataset.partitions.partition_names == set(partition_cols)
+        shutil.rmtree(path)
+
+    def test_ignore_partition_cols_lt_070(self, pa_lt_070, df_full):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        # supported in >= 0.7.0
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: Can remove this comment.

**Review by TomAugspurger — COMMENTED** [review]
We should mention somewhere that

File: pandas/tests/io/test_parquet.py (hunk 6 above)
Comment (anjsudh) [reply pandas/tests/io/test_parquet.py]: tm.ensure_clean() does not create Temporary Directory. It only creates temporary files. Will change the same in another request

Comment (gfyoung) [reply pandas/tests/io/test_parquet.py]: Silly question, but why are we passing in a directory for `df.to_parquet` ?

Comment (anjsudh) [reply pandas/tests/io/test_parquet.py]: We are passing the path for the temp directory. pyarrow.write_to_dataset requires a directory to store the partitioned files.

Comment (anjsudh) [reply pandas/tests/io/test_parquet.py]: @gfyoung For more details, do refer here : https://arrow.apache.org/docs/python/parquet.html#writing-to-partitioned-datasets

**Comment by anjsudh** [comment]
@WillAyd Hi, hope you can review the diff now? Have added the necessary tests

**Comment by xhochy** [comment]
From a `pyarrow` perspective this is LGTM.

File: pandas/io/parquet.py [hunk 10]
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
     path : string
-        File path
```
Comment (jreback) [inline pandas/io/parquet.py]: I am not sure this doc-string is formatted correctly. @datapythonista ?

also pls list the ``partition_cols`` arg with a versionadded tag (and provide a description).

this only applies to pyarrow? or also fastparquet?

File: pandas/tests/io/test_parquet.py (hunk 6 above)
Comment (jreback) [reply pandas/tests/io/test_parquet.py]: tempfile.mkdtemp is ok, but pls use this in a with statement so it is cleaned.

File: pandas/tests/io/test_parquet.py [hunk 11]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
...
+        path = tempfile.mkdtemp()
+        df.to_parquet(path, partition_cols=partition_cols,
+                      compression=None)
+        import pyarrow.parquet as pq
+        dataset = pq.ParquetDataset(path, validate_schema=False)
+        assert len(dataset.partitions.partition_names) == 2
+        assert dataset.partitions.partition_names == set(partition_cols)
+        shutil.rmtree(path)
+
+    def test_ignore_partition_cols_lt_070(self, pa_lt_070, df_full):
```
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: add the issue number as a comment

File: pandas/tests/io/test_parquet.py [hunk 12]
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
         check_round_trip(df_compat, pa,
                          path='s3://pandas-test/pyarrow.parquet')
 
+    def test_partition_cols_supported(self, pa_ge_070, df_full):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        path = tempfile.mkdtemp()
+        df.to_parquet(path, partition_cols=partition_cols,
+                      compression=None)
```
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: is there a reason you are also not reading this back with ``pd.read_parquet``?

**Review by jreback — CHANGES_REQUESTED** [review]
also pls update the documentatio in io.rst

File: pandas/io/parquet.py [hunk 13]
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
```
Comment (datapythonista) [inline pandas/io/parquet.py]: Can you remove this parameter please. We don't document the self.

File: pandas/io/parquet.py [hunk 14]
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
...
     path : string
```
Comment (datapythonista) [inline pandas/io/parquet.py]: Can you replace `string` by the Python type `str`?

File: pandas/io/parquet.py [hunk 15]
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
...
-        File path
+        File path ( Will be used as `root_path` if
+        `partition_cols` is provided as parameter for 'pyarrow' engine).
```
Comment (datapythonista) [inline pandas/io/parquet.py]: Can you rephrase this please. It's not very clear what you mean.

**Review by datapythonista — COMMENTED** [review]
Suggested couple of improvements to the docstring. Also, if you can run `./scripts/validate_docstrings.py pandas.DataFrame.to_parquet` to see that everything else is correct in it, that would be great.

File: pandas/tests/io/test_parquet.py (hunk 12 above)
Comment (anjsudh) [reply pandas/tests/io/test_parquet.py]: The objective of this test was to to ensure the file is partitioned. Hence covered that alone in this test. I am not clear on how to get the number of partitions in the file using read_parquet

File: pandas/io/parquet.py (hunk 10 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: this applies only for pyarrow

**Comment by anjsudh** [comment]
closes #23373 in a way

File: pandas/util/testing.py [hunk 16]
```diff
@@ -772,6 +772,23 @@ def ensure_clean(filename=None, return_filelike=False):
...
+    Yields
+    ----------
+    Temporary directory path
+    """
+    directory_name = tempfile.mkdtemp(suffix='')
+    try:
+        yield directory_name
+    finally:
+        import shutil
+        shutil.rmtree(directory_name)
```
Comment (TomAugspurger) [inline pandas/util/testing.py]: This should be in a `try / except Exception`, since we shouldn't ever fail in a test cleanup.

File: pandas/util/testing.py [hunk 17]
```diff
@@ -772,6 +772,23 @@ def ensure_clean(filename=None, return_filelike=False):
                 print("Exception on removing file: {error}".format(error=e))
 
 
+@contextmanager
+def ensure_clean_dir():
+    """
+    Get a temporary directory path and agrees to remove on close.
+
+    Yields
+    ----------
```
Comment (TomAugspurger) [inline pandas/util/testing.py]: Underline should match the length of Yields

File: pandas/io/parquet.py [hunk 18]
```diff
@@ -266,11 +278,18 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
         engine's default behavior will be used.
 
         .. versionadded 0.24.0
+    partition_cols : list, optional
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: Is it ignored, or doesn't it pass through to `write_table` and raise?

I would prefer that we raise for old pyarrows, so that a user doesn't think this is working when it isn't.

Also, I thought that fast parquet called it `partition_on`.

File: pandas/io/parquet.py [hunk 19]
```diff
@@ -224,7 +235,8 @@ def write(self, df, path, compression='snappy', index=None, **kwargs):
 
         with catch_warnings(record=True):
             self.api.write(path, df, compression=compression,
-                           write_index=index, **kwargs)
+                           write_index=index, partition_on=partition_cols,
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: What if a user does

```python
df.to_parquet("/tmp/a", engine="fastparquet", partition_on=['foo'])
```

I think that would raise an error, since you've specified `partition_on` twice. It should use the user-provided `partition_on`.

File: pandas/io/parquet.py (hunk 18 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: Yes, we will be passing this key as partition_on

File: pandas/io/parquet.py (hunk 19 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: i feel using one argument for partitioning makes more sense. Why don't we raise an error, asking them to use partition_cols instead in that case

File: pandas/core/frame.py [hunk 20]
```diff
@@ -1998,6 +1999,12 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
             the behavior depends on the chosen engine.
 
             .. versionadded:: 0.24.0
+        partition_cols : list, optional, default None
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
+            .. versionadded:: 0.24.0
```
Comment (jreback) [inline pandas/core/frame.py]: need a blank line before the versionadded. we should prob just bump our min of pyarrow to 0.8.0 anyhow, can you create a new issue for this.

File: pandas/io/parquet.py [hunk 21]
```diff
@@ -206,12 +213,16 @@ def __init__(self):
...
-    def write(self, df, path, compression='snappy', index=None, **kwargs):
+    def write(self, df, path, compression='snappy', index=None,
+              partition_cols=None, **kwargs):
         self.validate_dataframe(df)
         # thriftpy/protocol/compact.py:339:
         # DeprecationWarning: tostring() is deprecated.
         # Use tobytes() instead.
 
+        if partition_cols is not None:
+            kwargs['file_scheme'] = 'hive'
```
Comment (jreback) [inline pandas/io/parquet.py]: is this not defaulted pyarrow? if you don't specify what is the default?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parquet.py (hunk 21 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: this is for fastparquet. Default file_scheme is simple. In simple _metadata file will not be generated. _metadata file is necessary to perform a read on a directory in fastparquet

File: pandas/io/parquet.py (hunk 19 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: make sense.. will handle it such that partition_on gets more priority than partition_cols with a warning

Comment (TomAugspurger) [reply pandas/io/parquet.py]: I'm not sure it should even be a warning. If a warning is going to be emitted, then that should be done by fastparquet.

**Comment by TomAugspurger** [comment]
cc @martindurant for fastparquet things:

tldr: this PR is adding support for writing a dataframe partitioned by columns with the pyarrow engine.

1. fastparquet uses `partition_on`, pyarrow uses `partition_cols`. @xhochy @martindurant Any thoughts on standardizing on a name (stepping back, these do the same thing, correct?)
2. @martindurant thoughts on the change at 
https://github.com/pandas-dev/pandas/pull/23321#discussion_r228909418?

File: pandas/io/parquet.py (hunk 21 above)
Comment (martindurant) [reply pandas/io/parquet.py]: To be specific, simple means one file only, and is not appropriate for dask. There is also Drill scheme, where the directory names are different from the Hive variant (*if* using directory partitioning, no difference otherwise). Spark defaults to the Hive format, it's a reasonable default in every case.

Comment (martindurant) [reply pandas/io/parquet.py]: (the _metadata file is longer necessary for the fastparquet branch, actually, but it is preferred)

**Comment by anjsudh** [comment]
@WillAyd @jreback @TomAugspurger Have made the necessary changes. Hope you can have a look?

File: pandas/io/parquet.py [hunk 22]
```diff
@@ -206,12 +217,20 @@ def __init__(self):
...
-    def write(self, df, path, compression='snappy', index=None, **kwargs):
+    def write(self, df, path, compression='snappy', index=None,
+              partition_cols=None, **kwargs):
         self.validate_dataframe(df)
         # thriftpy/protocol/compact.py:339:
         # DeprecationWarning: tostring() is deprecated.
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            partition_cols = kwargs['partition_on']
```
Comment (jreback) [inline pandas/io/parquet.py]: use .pop()

File: pandas/io/parquet.py (hunk 21 above)
Comment (jreback) [reply pandas/io/parquet.py]: can you add a comment above this describing what you said above, e.g. for compat with fp

File: pandas/tests/io/test_parquet.py [hunk 23]
```diff
@@ -543,3 +565,27 @@ def test_s3_roundtrip(self, df_compat, s3_resource, fp):
...
+        df = df_full
+        with tm.ensure_clean_dir() as path:
+            df.to_parquet(path, engine="fastparquet",
+                          partition_cols=partition_cols, compression=None)
+            assert os.path.exists(path)
+            import fastparquet
+            actual_partition_cols = fastparquet.ParquetFile(path, False).cats
+            assert len(actual_partition_cols) == 2
+
+    def test_partition_on_supported(self, fp, df_full):
```
Comment (jreback) [inline pandas/tests/io/test_parquet.py]: what happens with a non-list-like partition column? does this need checking?

File: pandas/tests/util/test_testing.py [hunk 24]
```diff
@@ -875,3 +875,12 @@ def test_datapath_missing(datapath, request):
...
     assert result == expected
+
+
+def test_create_temp_directory():
+    temppath = ''
+    with tm.ensure_clean_dir() as path:
+        assert os.path.exists(path)
+        assert os.path.isdir(path)
+        temppath = path
+    assert not os.path.exists(temppath)
```
Comment (jreback) [inline pandas/tests/util/test_testing.py]: you can just use path here, its still in scope

File: pandas/util/testing.py [hunk 25]
```diff
@@ -772,6 +772,26 @@ def ensure_clean(filename=None, return_filelike=False):
...
+    Yields
+    ------
+    Temporary directory path
+    """
+    directory_name = tempfile.mkdtemp(suffix='')
+    try:
+        yield directory_name
+    finally:
+        try:
+            import shutil
```
Comment (jreback) [inline pandas/util/testing.py]: this import can be at the top of the file

File: pandas/util/testing.py [hunk 26]
```diff
@@ -772,6 +772,26 @@ def ensure_clean(filename=None, return_filelike=False):
...
+            shutil.rmtree(directory_name)
+        except Exception as e:
+            print("Exception on removing folder: {error}".format(error=e))
```
Comment (jreback) [inline pandas/util/testing.py]: just pass

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_parquet.py (hunk 23 above)
Comment (anjsudh) [reply pandas/tests/io/test_parquet.py]: both fastparquet and pyarrow require a list as input. So even if only one column is required, it is given as a list.

File: doc/source/io.rst [hunk 27]
```diff
@@ -4574,6 +4574,8 @@ Several caveats.
 * Categorical dtypes can be serialized to parquet, but will de-serialize as ``object`` dtype.
 * Non supported types include ``Period`` and actual Python object types. These will raise a helpful error message
   on an attempt at serialization.
+* ``partition_cols`` will be used for partitioning the dataset, where the dataset will be written to multiple
```
Comment (TomAugspurger) [inline doc/source/io.rst]: The rest of the items in this lists feel more like limitations of pandas / these engines. Requiring that `path` be a directory when `partition_cols` is set doesn't seem to fit here.

I think this is important / different enough to deserve a new small section below "Handling Indexes", with

1. A description of what `partition_cols` requires (list of column names, directory for file path)
2. A description of why you might want to use partition_cols
3. A small example.

File: doc/source/whatsnew/v0.24.0.txt [hunk 28]
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- :func:`~DataFrame.to_parquet` now supports writing a DataFrame as a directory of parquet files partitioned by a subset of the columns. (:issue:`23283`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: Probably best to mention "with the pyarrow engine (this was previously supported with fastparquet)."

File: pandas/core/frame.py [hunk 29]
```diff
@@ -1997,6 +1998,13 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
             If ``False``, they will not be written to the file. If ``None``,
             the behavior depends on the chosen engine.
 
+            .. versionadded:: 0.24.0
+        partition_cols : list, optional, default None
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Is it actually ignored for older pyarrows? I would have hoped it would raise when pyarrow gets the unrecognized argument.

Comment (TomAugspurger) [reply pandas/core/frame.py]: Actually it seems like we raise. Could you update this?

File: pandas/io/parquet.py [hunk 30]
```diff
@@ -208,12 +219,19 @@ def __init__(self):
...
-    def write(self, df, path, compression='snappy', index=None, **kwargs):
+    def write(self, df, path, compression='snappy', index=None,
+              partition_cols=None, **kwargs):
         self.validate_dataframe(df)
         # thriftpy/protocol/compact.py:339:
         # DeprecationWarning: tostring() is deprecated.
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            partition_cols = kwargs.pop('partition_on')
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: Should check that `partition_cols` is None before doing this. We would want to raise on

```python
df.to_parquet(engine='fastparquet', partition_cols=['a', 'b'], partition_on=['b', 'c'])
```

File: pandas/io/parquet.py (hunk 21 above)
Comment (TomAugspurger) [reply pandas/io/parquet.py]: Need a comment here sayings what's going on.

File: pandas/io/parquet.py [hunk 31]
```diff
@@ -268,11 +287,18 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
         engine's default behavior will be used.
 
         .. versionadded 0.24.0
+    partition_cols : list, optional
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Comment (TomAugspurger) [inline pandas/io/parquet.py]: Update this: We don't ignore.

File: pandas/io/parquet.py (hunk 21 above)
Comment (martindurant) [reply pandas/io/parquet.py]: That is expected in 0.1.6, you would need to pass the whole list of data files rather than the root directory. This may be fixed in master.

File: pandas/io/parquet.py (hunk 31 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: missed updating the docs.. thanks :)

File: pandas/io/parquet.py (hunk 30 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: done

File: pandas/core/frame.py (hunk 29 above)
Comment (anjsudh) [reply pandas/core/frame.py]: done

File: doc/source/whatsnew/v0.24.0.txt (hunk 28 above)
Comment (anjsudh) [reply doc/source/whatsnew/v0.24.0.txt]: done

File: doc/source/io.rst (hunk 27 above)
Comment (anjsudh) [reply doc/source/io.rst]: done

**Comment by TomAugspurger** [comment]
Merge conflict in `io/parquet.py`, if you could fix that up I think this will be good to go.

File: pandas/io/parquet.py [hunk 32]
```diff
@@ -194,15 +213,15 @@ def read(self, path, columns=None, **kwargs):
...
     Write a DataFrame to the parquet format.
 
     Parameters
     ----------
-    df : DataFrame
-    path : string
-        File path
+    path : str
+        File path or Root Directory path. Will be used as Root Directory path
+        while writing a partitioned dataset.
```
Comment (jreback) [inline pandas/io/parquet.py]: versionchanged tag here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parquet.py (hunk 32 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: done

File: pandas/core/frame.py [hunk 33]
```diff
@@ -1984,7 +1984,8 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
         Parameters
         ----------
         fname : str
-            String file path.
+            File path or Root Directory path. Will be used as Root Directory
+            path while writing a partitioned dataset.
```
[the remark this answers is not in the document]
Comment (anjsudh) [reply pandas/core/frame.py]: done

File: doc/source/whatsnew/v0.24.0.txt [hunk 34]
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- With the pyarrow engine, :func:`~DataFrame.to_parquet` now supports writing a DataFrame as a directory of parquet files partitioned by a subset of the columns. (:issue:`23283`).
```
[the remark this answers is not in the document]
Comment (anjsudh) [reply doc/source/whatsnew/v0.24.0.txt]: done

File: doc/source/io.rst [hunk 35]
```diff
@@ -4668,6 +4668,33 @@ Passing ``index=True`` will *always* write the index, even if that's not the
...
+.. ipython:: python
+
+    df = pd.DataFrame({'a': [0, 0, 1, 1], 'b': [0, 1, 0, 1]})
+    df.to_parquet(fname='test', engine='pyarrow', partition_cols=['a'], compression=None)
+
+The `fname` specifies the parent directory to which data will be saved.
+The `partition_cols` are the column names by which the dataset will be partitioned.
+Columns are partitioned in the order they are given. The partition splits are
+determined by the unique values in the partition columns.
+The above example creates a partitioned dataset that may look like:
```
[the remark this answers is not in the document]
Comment (anjsudh) [reply doc/source/io.rst]: done

File: doc/source/io.rst [hunk 36]
```diff
@@ -4668,6 +4668,33 @@ Passing ``index=True`` will *always* write the index, even if that's not the
 underlying engine's default behavior.
 
 
+Partitioning Parquet files
+''''''''''''''''''''''''''
```
[the remark this answers is not in the document]
Comment (anjsudh) [reply doc/source/io.rst]: done

File: doc/source/io.rst [hunk 37]
```diff
@@ -4668,6 +4668,36 @@ Passing ``index=True`` will *always* write the index, even if that's not the
...
+
+.. versionadded:: 0.24.0
+
+Parquet supports partitioning of data based on the values of one or more columns.
+
+.. ipython:: python
+   :suppress:
```
Comment (jreback) [inline doc/source/io.rst]: don't add the suppress here (this won't print the block). Rather add another block after than cleans up the dir that is created (and use suppress on that)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by TomAugspurger — APPROVED** [review]

**Comment by anjsudh** [comment]
@jreback @WillAyd hope you can have a look ?

File: pandas/core/frame.py [hunk 38]
```diff
@@ -1984,7 +1984,10 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
         Parameters
         ----------
         fname : str
```
Comment (jreback) [inline pandas/core/frame.py]: side issue. we use ``path`` elsewhere for IO routines. We should change this as well (out of scope here). would have to deprecate (the name) unfortunately.

Comment (jreback) [reply pandas/core/frame.py]: we actually use path on the top-level ``.to_parquet``, not sure how this is named this way.

**Review by jreback — APPROVED** [review]
off topic comments, lgtm. @WillAyd

File: doc/source/io.rst [hunk 39]
```diff
@@ -4668,6 +4668,43 @@ Passing ``index=True`` will *always* write the index, even if that's not the
...
+        ├── e6ab24a4f45147b49b54a662f0c412a3.parquet
+        └── ...
+
+.. ipython:: python
+   :suppress:
+
+   from shutil import rmtree
+   try:
+       rmtree('test')
+   except Exception:
```
Comment (WillAyd) [inline doc/source/io.rst]: Can we be more specific about the exception here? I believe this could just be an `OSError` or `FileNotFoundError` depending on version. Would rather catch those so that this doesn't end up catching things it doesn't mean to during the doc build process

File: doc/source/whatsnew/v0.24.0.txt [hunk 40]
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- :func:`~DataFrame.to_parquet` now supports writing a ``DataFrame`` as a directory of parquet files partitioned by a subset of the columns, when ``engine = 'pyarrow'``. (:issue:`23283`).
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.txt]: Very minor nits but you don't need the comma after "subset of the columns" nor do you need the two periods towards the end

File: pandas/core/frame.py [hunk 41]
```diff
@@ -1984,7 +1984,10 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
...
-            String file path.
+            File path or Root Directory path. Will be used as Root Directory
+            path while writing a partitioned dataset.
+
+            .. versionchanged:: 0.24.0
```
Comment (WillAyd) [inline pandas/core/frame.py]: Does this render? Might need a blank space after the directive?

File: pandas/core/frame.py [hunk 42]
```diff
@@ -1997,6 +2000,11 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
             If ``False``, they will not be written to the file. If ``None``,
             the behavior depends on the chosen engine.
 
+            .. versionadded:: 0.24.0
```
Comment (WillAyd) [inline pandas/core/frame.py]: Same question as above

File: pandas/io/parquet.py [hunk 43]
```diff
@@ -156,12 +162,24 @@ def __init__(self):
...
-    def write(self, df, path, compression='snappy', index=None, **kwargs):
+    def write(self, df, path, compression='snappy', index=None,
+              partition_cols=None, **kwargs):
         self.validate_dataframe(df)
         # thriftpy/protocol/compact.py:339:
         # DeprecationWarning: tostring() is deprecated.
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            if partition_cols is None:
```
Comment (WillAyd) [inline pandas/io/parquet.py]: Let's simplify this as `if 'partition_on' in kwargs and partition_cols is None`

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/io.rst (hunk 39 above)
Comment (WillAyd) [reply doc/source/io.rst]: Hmm OK. No concern around it catching exceptions that it shouldn't though?

**Comment by TomAugspurger** [comment]
Nope. I don't really care about bugs in shutil :)

File: doc/source/whatsnew/v0.24.0.txt (hunk 40 above)
Comment (anjsudh) [reply doc/source/whatsnew/v0.24.0.txt]: Done

File: pandas/io/parquet.py (hunk 43 above)
Comment (anjsudh) [reply pandas/io/parquet.py]: Done

File: pandas/core/frame.py (hunk 41 above)
Comment (anjsudh) [reply pandas/core/frame.py]: Looks like it does not.

Documentation shows API Reference for DataFrame.to_parquet, read_parquet() for parquet.py and io.rst, which explains about reading and writing to parquet.

Comment (WillAyd) [reply pandas/core/frame.py]: Wasn't referring to the public API as much as the docstring in isolation. If you can ensure there is a blank line after this and any of the other `versionadded` directives should be the last thing

Comment (anjsudh) [reply pandas/core/frame.py]: have added an empty line after all versionchanged and versionadded. Hope you can have a look

**Comment by TomAugspurger** [comment]
Looks good. Nice work @anjsudh!
